File: src/crypto/crypto_manager.cpp
#include "crypto_manager.hpp"
#include <openssl/pem.h>
#include <openssl/x509.h>
#include <fstream>
#include <sstream>
#include <iostream>
#include <filesystem>
#include <vector>

namespace nemo
{

    CryptoManager::CryptoManager() {}
    CryptoManager::~CryptoManager() {}

    bool CryptoManager::initialize()
    {
        return loadKeysAndCertificateFromFile();
    }

    std::string CryptoManager::getCertificate() const { return certificate_; }
    std::string CryptoManager::getPrivateKey() const { return private_key_; }

    // Cerca il file nei path standard (stessa logica di aasdk::Cryptor)
    static std::string findFile(const std::vector<std::string> &paths)
    {
        for (const auto &p : paths)
        {
            if (std::filesystem::exists(p))
            {
                std::ifstream f(p);
                if (f.good())
                {
                    std::ostringstream ss;
                    ss << f.rdbuf();
                    std::cout << "[CryptoManager] Loaded from: " << p << std::endl;
                    return ss.str();
                }
            }
        }
        return {};
    }

    bool CryptoManager::loadKeysAndCertificateFromFile()
    {
        // Stessi path di aasdk::Cryptor (vedi log di avvio)
        const std::vector<std::string> certPaths = {
            "/etc/openauto/headunit.crt",
            "/usr/share/aasdk/cert/headunit.crt",
            "./cert/headunit.crt"};
        const std::vector<std::string> keyPaths = {
            "/etc/openauto/headunit.key",
            "/usr/share/aasdk/cert/headunit.key",
            "./cert/headunit.key"};

        certificate_ = findFile(certPaths);
        if (certificate_.empty())
        {
            std::cerr << "[CryptoManager] ERRORE: nessun certificato trovato!" << std::endl;
            return false;
        }

        private_key_ = findFile(keyPaths);
        if (private_key_.empty())
        {
            std::cerr << "[CryptoManager] ERRORE: nessuna chiave privata trovata!" << std::endl;
            return false;
        }

        // Validazione rapida: cert e chiave devono corrispondersi
        BIO *certBio = BIO_new_mem_buf(certificate_.c_str(), -1);
        X509 *x509 = PEM_read_bio_X509(certBio, nullptr, nullptr, nullptr);
        BIO_free(certBio);

        BIO *keyBio = BIO_new_mem_buf(private_key_.c_str(), -1);
        EVP_PKEY *pk = PEM_read_bio_PrivateKey(keyBio, nullptr, nullptr, nullptr);
        BIO_free(keyBio);

        if (!x509 || !pk)
        {
            std::cerr << "[CryptoManager] ERRORE: parsing PEM fallito!" << std::endl;
            if (x509)
                X509_free(x509);
            if (pk)
                EVP_PKEY_free(pk);
            return false;
        }

        // Verifica che la chiave corrisponda al certificato
        if (X509_check_private_key(x509, pk) != 1)
        {
            std::cerr << "[CryptoManager] ERRORE: chiave e certificato NON corrispondono!" << std::endl;
            X509_free(x509);
            EVP_PKEY_free(pk);
            return false;
        }

        std::cout << "[CryptoManager] Certificato e chiave validati con successo." << std::endl;

        X509_free(x509);
        EVP_PKEY_free(pk);
        return true;
    }

} // namespace nemo

File: python/app/crypto_manager.py
"""
python/app/crypto_manager.py

Equivalente Python di src/crypto/crypto_manager.cpp.

Responsabilità:
  - Ricerca del certificato TLS e della chiave privata sui path standard
  - Validazione PEM (parsing X.509 + verifica corrispondenza chiave pubblica)
  - Nessun binding C++ richiesto: usa la libreria `cryptography` (PyCA)

Dipendenze:
  pip install cryptography

Path cercati (stesso ordine di aasdk::Cryptor):
  Certificato : /etc/openauto/headunit.crt
                /usr/share/aasdk/cert/headunit.crt
                ./cert/headunit.crt
  Chiave priv.: /etc/openauto/headunit.key
                /usr/share/aasdk/cert/headunit.key
                ./cert/headunit.key

Uso:
    crypto = CryptoManager()
    if not crypto.initialize():
        raise RuntimeError("Certificati non trovati o non validi")
    usb.set_certificate_and_key(crypto.get_certificate(), crypto.get_private_key())
"""

from __future__ import annotations

import os
import logging
from typing import Optional

log = logging.getLogger(__name__)

# Percorsi standard (allineati a aasdk::Cryptor e crypto_manager.cpp)
_CERT_PATHS: list[str] = [
    "/etc/openauto/headunit.crt",
    "/usr/share/aasdk/cert/headunit.crt",
    "./cert/headunit.crt",
]

_KEY_PATHS: list[str] = [
    "/etc/openauto/headunit.key",
    "/usr/share/aasdk/cert/headunit.key",
    "./cert/headunit.key",
]


def _find_file(paths: list[str]) -> Optional[str]:
    """Cerca il primo file esistente e leggibile tra i path forniti."""
    for p in paths:
        if os.path.isfile(p):
            try:
                with open(p, "r") as f:
                    content = f.read()
                log.info("[CryptoManager] Caricato: %s", p)
                return content
            except OSError as e:
                log.warning("[CryptoManager] Impossibile leggere %s: %s", p, e)
    return None


def _validate_cert_key_pair(cert_pem: str, key_pem: str) -> bool:
    """
    Verifica che la chiave privata corrisponda al certificato X.509.
    Equivalente di X509_check_private_key() usato in crypto_manager.cpp.

    Usa `cryptography` (PyCA) — nessun binding C++/OpenSSL diretto.
    """
    try:
        from cryptography import x509
        from cryptography.hazmat.primitives.serialization import (
            load_pem_private_key,
            Encoding,
            PublicFormat,
        )
        from cryptography.hazmat.backends import default_backend

        backend = default_backend()

        cert = x509.load_pem_x509_certificate(cert_pem.encode(), backend)
        key  = load_pem_private_key(key_pem.encode(), password=None, backend=backend)

        cert_pub_bytes = cert.public_key().public_bytes(
            Encoding.PEM, PublicFormat.SubjectPublicKeyInfo
        )
        key_pub_bytes = key.public_key().public_bytes(
            Encoding.PEM, PublicFormat.SubjectPublicKeyInfo
        )

        if cert_pub_bytes != key_pub_bytes:
            log.error("[CryptoManager] ERRORE: chiave e certificato NON corrispondono!")
            return False

        log.info("[CryptoManager] Certificato e chiave validati con successo.")
        return True

    except ImportError:
        # cryptography non installata: skip validazione, solo warning
        log.warning(
            "[CryptoManager] libreria `cryptography` non trovata: "
            "validazione PEM saltata. Installa con: pip install cryptography"
        )
        return True  # non bloccare il boot se la lib non c'è

    except Exception as e:
        log.error("[CryptoManager] ERRORE parsing PEM: %s", e)
        return False


class CryptoManager:
    """
    Carica e valida il certificato TLS e la chiave privata per aasdk.

    Equivalente Python completo di src/crypto/crypto_manager.cpp.
    Non richiede compilazione C++.

    Esempio d'uso:
        crypto = CryptoManager()
        if not crypto.initialize():
            sys.exit(1)
        usb.set_certificate_and_key(
            crypto.get_certificate(),
            crypto.get_private_key()
        )
    """

    def __init__(
        self,
        cert_paths: Optional[list[str]] = None,
        key_paths:  Optional[list[str]] = None,
    ):
        """
        Parametri opzionali per override dei path di default.
        Se non forniti, usa i path standard di aasdk.
        """
        self._cert_paths: list[str] = cert_paths or _CERT_PATHS
        self._key_paths:  list[str] = key_paths  or _KEY_PATHS
        self._certificate: str = ""
        self._private_key: str = ""

    # ------------------------------------------------------------------
    # API pubblica — stessa interfaccia di CryptoManager C++
    # ------------------------------------------------------------------

    def initialize(self) -> bool:
        """Carica e valida cert + key. Ritorna True se tutto ok."""
        cert_pem = _find_file(self._cert_paths)
        if not cert_pem:
            log.error(
                "[CryptoManager] ERRORE: nessun certificato trovato nei path: %s",
                self._cert_paths,
            )
            return False

        key_pem = _find_file(self._key_paths)
        if not key_pem:
            log.error(
                "[CryptoManager] ERRORE: nessuna chiave privata trovata nei path: %s",
                self._key_paths,
            )
            return False

        if not _validate_cert_key_pair(cert_pem, key_pem):
            return False

        self._certificate = cert_pem
        self._private_key = key_pem
        return True

    def get_certificate(self) -> str:
        """Ritorna il certificato PEM come stringa."""
        return self._certificate

    def get_private_key(self) -> str:
        """Ritorna la chiave privata PEM come stringa."""
        return self._private_key
